src/vsr/checkpoint_trailer.zig

```diff
@@ -140,269 +143,230 @@ pub fn FreeSetEncodedType(comptime Storage: type) type {
-            reference: vsr.SuperBlockFreeSetReference,
-            callback: *const fn (set: *Self) void,
+            reference: vsr.SuperBlockTrailerReference,
+            callback: *const fn (trailer: *Self) void,
         ) void {
-            set.grid = grid;
-            assert(!set.grid.?.free_set.opened);
+            trailer.grid = grid;
 
-            assert(set.callback == .none);
-            defer assert(set.callback == .open);
+            assert(trailer.callback == .none);
+            defer assert(trailer.callback == .open);
 
-            assert(reference.free_set_size % @sizeOf(FreeSet.Word) == 0);
-            assert(set.size == 0);
-            assert(set.size_transferred == 0);
-            assert(set.block_index == 0);
+            assert(reference.trailer_size % @sizeOf(FreeSet.Word) == 0);
+            assert(trailer.size == 0);
+            assert(trailer.size_transferred == 0);
+            assert(trailer.block_index == 0);
 
-            set.size = reference.free_set_size;
-            set.checksum = reference.checksum;
-            set.callback = .{ .open = callback };
+            trailer.size = reference.trailer_size;
+            trailer.checksum = reference.checksum;
+            trailer.callback = .{ .open = callback };
 
             // Start from the last block, as the linked list arranges data in the reverse order.
-            set.block_index = set.block_count();
+            trailer.block_index = trailer.block_count();
 
-            if (set.size == 0) {
+            if (trailer.size == 0) {
                 assert(reference.last_block_address == 0);
-                set.grid.?.on_next_tick(open_next_tick, &set.next_tick);
+                trailer.grid.?.on_next_tick(open_next_tick, &trailer.next_tick);
             } else {
                 assert(reference.last_block_address != 0);
-                set.open_read_next(reference.last_block_address, reference.last_block_checksum);
+                trailer.open_read_next(reference.last_block_address, reference.last_block_checksum);
             }
         }
 
         fn open_next_tick(next_tick: *Grid.NextTick) void {
-            const set = @fieldParentPtr(Self, "next_tick", next_tick);
-            assert(set.callback == .open);
-            assert(set.size == 0);
-            set.open_done();
+            const trailer = @fieldParentPtr(Self, "next_tick", next_tick);
+            assert(trailer.callback == .open);
+            assert(trailer.size == 0);
+            trailer.open_done();
         }
 
-        fn open_read_next(set: *Self, address: u64, checksum: u128) void {
-            assert(set.callback == .open);
-            assert(set.size > 0);
+        fn open_read_next(trailer: *Self, address: u64, checksum: u128) void {
+            assert(trailer.callback == .open);
+            assert(trailer.size > 0);
+            assert((trailer.size_transferred == 0) ==
+                (trailer.block_index == trailer.block_count()));
```

Comment:
:+1: for grouping related assert together. Do we have a rule of thumb which asserts go first, those that check the state of `self`, or those that check validity of parameters?

Reply:
My personal rule of thumb is to put `self` assertions first, since `self` is the first parameter.

Reply:
:facepalm: that's genius!